pkg/identity/kubernetes/principal.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"context"
+	"crypto/x509"
+	"net/url"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/sigstore/fulcio/pkg/ca/x509ca"
+	"github.com/sigstore/fulcio/pkg/identity"
+)
+
+type principal struct {
+	// Subject ('sub') from ID token
+	subject string
+
+	// Issuer ('iss') from ID token
+	issuer string
+
+	// URI to be set in certificate. URI is of the form
+	// https://kubernetes.io/namespaces/<namespace>/serviceaccounts/<serviceaccount>.
+	uri string
+}
+
+func PrincipalFromIDToken(ctx context.Context, token *oidc.IDToken) (identity.Principal, error) {
+	k8sURI, err := kubernetesToken(token)
+	if err != nil {
+		return nil, err
+	}
+	return principal{
+		subject: token.Subject,
+		issuer:  token.Issuer,
+		uri:     k8sURI,
+	}, nil
+}
+
+func (p principal) Name(context.Context) string {
```

Comment:
Sorry, just noticed this - This (and below) should be a pointer receiver, to avoid the copy of the principal. Even though the receiver is not mutated, it saves a copy.

Reply:
I think that communicating to the reader that the values are immutable is more important that any performance overhead here.

That said, if performance is a concern I've benchmarked both here: https://github.com/nsmith5/receiverbench. The allocations are the same in each case and the speed looks identitical if not slightly faster for the value receiver. For such a simply example its makes sense to me that the compiler would be able to optimize away any difference between these two.

Reply:
It’s likely a small difference because the struct is a small size. At scale, handling 1000s of QPS, this impact can become more significant with more allocations.

There’s no official style guide, the best I’ve found is https://github.com/golang/go/wiki/CodeReviewComments#receiver-type. 
The guidance here is prefer a pointer unless there’s a good reason not to, and be consistent. Generally across our code base, I see pointer receivers preferred. 

I agree that there’s a benefit to clearly showing it’s immutable, but I think it’s also pretty clear from the function that the method doesn’t mutate the struct fields, given they’re small functions.

Reply:
I don't feel incredibly strongly about this, I'd just prefer consistency

Reply:
Thought more, this is fine, it’s consistent within the principal package